### PR TITLE
Minor updates and code cleanups

### PR DESCRIPTION
**Pull request by balbirthomas**
- Updated Library overview docs to reflect recent changes made to the interface
- Set unit names in relation data
- Updated unit tests to reflect unit names are present in relation data
- Correct docstrings of individual methods

Deferred to future PRs.
- Review of test cases and perhaps add more unit tests
- Other code cleanups (such as refactor of rules provider and related objects for which github tickets have been created).

### CODE REVIEW

**Comment by sed-i** [comment]
Might be worthwhile to throw [codespell](https://github.com/canonical/operator-template/pull/9/files) into tox lint env?

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 1]
```diff
@@ -293,10 +287,9 @@ def _on_scrape_targets_changed(self, event):
 (https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config).
 
 Units of consumer charm advertise their address over unit relation
-data using the `prometheus_scrape_host` key. While the
+data using the `prometheus_scrape_host_address` key. While the
```
Comment (mmanciop) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:290]: @balbirthomas please remind me why we did this? It's going to come _really_ in handy with `ingress`, but I am not sure we had that in mind at the time.

File: tests/unit/test_endpoint_consumer.py [hunk 2]
```diff
@@ -179,7 +184,12 @@ def setup_charm_relations(self, multi=False):
             )
             self.harness.add_relation_unit(rel_id, "other-consumer/0")
             self.harness.update_relation_data(
-                rel_id, "other-consumer/0", {"prometheus_scrape_host": "2.2.2.2"}
+                rel_id,
+                "other-consumer/0",
+                {
+                    "prometheus_scrape_host_address": "2.2.2.2",
+                    "prometheus_scrape_host_name": "other-consumer/0",
```
Comment (mmanciop) [inline tests/unit/test_endpoint_consumer.py:191]: Bad name, it is a `unit_name`, not a `host_name`

**Review by mmanciop — CHANGES_REQUESTED** [review]
I do not agree with delivering the unit name over a field called `host_name`. It is going to cause a lot of WTF all around :-)

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 1 above)
Comment (balbirthomas) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:290]: When the Prometheus job specification use wildcard hosts (`*:8080`) then `MetricsEndpointConsumer` needs the IP addresses of all the units. There were two options at that time 1) `MetricsEndpointConsumer` should gather the IP address from relation event data 2) `MetricsEndpointProvider` units forward this data in unit relation data. I am not sure if I remember the precise details why we went down route 2 but I think at that time there was lot of discussion on how this should be done. The consensus in the team was for the provider charm in this case to forward the IP address through unit relation data. If I remember correctly this was the pattern Jeremy was recommending. I do remember creating a [ticket](https://github.com/canonical/operator/issues/560) for `ops` also. I think Graylog charm (at least Jeremy's branch) and other charms may also be doing the same. Happy to investigate if you think this needs changing.

File: tests/unit/test_endpoint_consumer.py (hunk 2 above)
Comment (balbirthomas) [reply tests/unit/test_endpoint_consumer.py:191]: Ok will change.

**Comment by balbirthomas** [comment]
Manual test of unit names in cross model relations.
Screenshot from 2021-12-02 14-31-13

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 3]
```diff
@@ -953,11 +941,14 @@ def _relation_hosts(self, relation) -> dict:
...
         """
-        return {
-            unit.name: relation.data[unit].get("prometheus_scrape_host")
-            for unit in relation.units
-            if relation.data[unit].get("prometheus_scrape_host")
-        }
+        hosts = {}
+        for unit in relation.units:
+            unit_name = relation.data[unit].get("prometheus_scrape_unit_name")
+            host_address = relation.data[unit].get("prometheus_scrape_unit_address")
```
Comment (mmanciop) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:947]: If we do a fallback to the previous label name we do not break compat with the previous versions of this lib.
Also, the name of the variable is wrong :-)

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 4]
```diff
@@ -953,11 +941,14 @@ def _relation_hosts(self, relation) -> dict:
...
             the specified relation.
         """
-        return {
-            unit.name: relation.data[unit].get("prometheus_scrape_host")
-            for unit in relation.units
-            if relation.data[unit].get("prometheus_scrape_host")
-        }
+        hosts = {}
+        for unit in relation.units:
+            unit_name = relation.data[unit].get("prometheus_scrape_unit_name")
```
Comment (mmanciop) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:946]: This is not handling the case in which the other side is not adding the unit name, which is the case for every charm out there atm :-)

I feel we are being a touch too cavalier about backwards compatibility here.

**Review by mmanciop — CHANGES_REQUESTED** [review]
Getting there, but there is no good reason to break backwards compatibility like this, even in a v0 version :-)

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 3 above)
Comment (balbirthomas) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:947]: You mean leave the name as `prometheus_scrape_host` ?

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 4 above)
Comment (balbirthomas) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:946]: This change does break compatibility with previous version. Charms will need to updated to the current version of the charm library. If ùnit_name` is not set then that scrape target is not scrapped. You mean we should allow scrapping of charms that do not set unit names ? If so in this case should the unit names be taken from the event as was previously done ?

Comment (mmanciop) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:946]: I mean that, if the consumer is talking with an older version of the producer that sends relation data, it should not break. In such a case, keeping the old behavior as fallback is fine.

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 3 above)
Comment (balbirthomas) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:947]: OK. I presume we do intend to deprecate `prometheus_scrape_host` at some point. I will put a TODO to this effect.

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 4 above)
Comment (balbirthomas) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py]: Done.

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 3 above)
Comment (balbirthomas) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py]: Done.

**Review by mmanciop — APPROVED** [review]

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 5]
```diff
@@ -1142,7 +1240,7 @@ def _resolve_dir_against_charm_path(charm: CharmBase, *path_elements: str) -> st
 
 
 class MetricsEndpointProvider(Object):
-    """Construct a metrics provider for a Prometheus charm."""
+    """An metrics endpoint for Prometheus."""
```
Comment (simskij) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:1243]:
```suggestion
    """A metrics endpoint for Prometheus."""
```

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 6]
```diff
@@ -1155,16 +1253,20 @@ def __init__(
 
         If your charm exposes a Prometheus metrics endpoint, the
         `MetricsEndpointProvider` object enables your charm to easily
-        communicate to a consumer how to reach that metrics endpoint.
+        communicate how to reach that metrics endpoint.
 
-        A charm instantiating this object has the metrics endpoints
-        of each of its units scraped by the related Prometheus
+        By default a charm instantiating this object has the metrics
```
Comment (simskij) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:1258]:
```suggestion
        By default, a charm instantiating this object has the metrics
```

**Review by simskij — CHANGES_REQUESTED** [review]

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 5 above)
Comment (balbirthomas) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py]: done.

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 6 above)
Comment (balbirthomas) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py]: done.

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 7]
```diff
@@ -1359,15 +1462,20 @@ def _scrape_metadata(self) -> dict:
 
 
 class PrometheusRulesProvider(Object):
-    """A 'prometheus_scrape' provider class for rule files only.
+    """Forward only rules to Prometheus.
```
[the remark this answers is not in the document]
Comment (balbirthomas) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py]: done.

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 8]
```diff
@@ -1361,16 +1461,21 @@ def _scrape_metadata(self) -> dict:
...
-    """A 'prometheus_scrape' provider class for rule files only.
+class PrometheusRulesProvider(Object):
+    """Forward rules to Prometheus.
 
-    This class is for sending rules to prometheus, unlike :class:`MetricsEndpointProvider`, which
-    is used for configuring a scrape target (including rules).
-    This is useful for providing "free standing" rules.
+    This object may be used to forward rules to Prometheus. At present it only supports
+    forwarding alert rules. This is unlike :class:`MetricsEndpointProvider`, which
+    is used for forwarding both scrape targets and associated alert rules. This object
```
Comment (sed-i) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:1469]:
```suggestion
    forwarding alert rules (support for recording rules is not yet implemented).
    In contrast to :class:`MetricsEndpointProvider`, which is used for forwarding both
    scrape job and associated alert rules, this object is only for forwarding 
    free-standing rules. This object
```

Comment (balbirthomas) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:1469]: I am fine with most of this suggestion with the only minor exception of the use of the phrase "free-standing rules". I would rather not invent such phraseology unless it is coming from a product specification. Are you ok with that ?

**Review by mmanciop — DISMISSED** [review]

**Review by simskij — APPROVED** [review]

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 8 above)
Comment (simskij) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:1469]: I agree with you on this @balbirthomas. The less internal nomenclature we introduce, the better - at least from an onboarding perspective. In that case, maybe changing `free-standing rules` to `rules from the lma-rules charm` would make sense?

Comment (mmanciop) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:1469]: Well we are going to rename the `lma-rules` charm, sooo... :D

**Review by mmanciop — APPROVED** [review]

**Pull request merged**